File: leetCode/101symmetric-tree/solution.java
import java.util.ArrayDeque;
import java.util.Deque;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        if(root == null)return true;
        return dfs(root.left, root.right);  
    }

    public boolean dfs(TreeNode leftTree,TreeNode rightTree){
        if(leftTree == null && rightTree == null){
            return true;
        }
        if(leftTree == null || rightTree == null){
            return false;
        }
        return leftTree.val == rightTree.val && dfs(leftTree.left,rightTree.right) && dfs(leftTree.right, rightTree.left);

        // if(leftTree == null && rightTree == null){
        //     return true;
        // }
        // else if(leftTree == null && rightTree != null){
        //     return false;
        // }
        // else if(leftTree != null && rightTree == null){
        //     return false;
        // }
        // else{
        //     if(leftTree.val != rightTree.val)return false;
        //     if(!dfs(leftTree.left,rightTree.right))return false;
        //     if(!dfs(leftTree.right, rightTree.left))return false;
        //     return true;
        // }
    }

    public class TreeNode {
        int val;                                                                                                                                                                     
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }
} 
















































































































































         

File: leetCode/101symmetric-tree/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function(root) {
    if(root == null) return true;
    const { left,right } = root;
    return dfs(left,right);

    function dfs(p, q){
        if(p == null && q == null) return true;
        if(p == null || q == null) return false;
        return p.val == q.val && dfs(p.left, q.right) && dfs(p.right, q.left);
    }
};